### PR TITLE
Add ormqr primitive and jax.scipy.linalg.qr_multiply

### PR DESCRIPTION
**Pull request by jpbrodrick89**
This addresses some of #29173 building the foundation for more efficient least squares solves, for example

```
  @jax.jit
  def qr_multiply_lstsq(A, b):
      Qtb, R = jax.scipy.linalg.qr_multiply(A, b, mode='right')
      return jax.scipy.linalg.solve_triangular(R, Qtb)
```

is 2-4x faster than the standard least square SVD solves and about 1.7x faster than using full Q materialisation.

A key limitation is that I have not defined  a jvp rule or primitive for `qr_multiply` as this is quite complex (and possibly MORE complex than a general jvp rule for the common least squares case, which I'm working on separately). We could either document this limitation or just remove the scipy API entirely and just keep the lax ormqr to be used directly as a building block in least squares solve. For example we could use this directly in lineax without waiting for further features in jax least squares.

The fallback approach applies sequential householder vectors with a fori_loop rather than adding the complex LAPACK blocking functionality.

CI will likely fail as this requires a new build of jaxlib. I rebuilt jaxlib from source and all linalg tests pass.

### CODE REVIEW

File: jax/_src/lax/linalg.py [hunk 1]
```diff
@@ -1394,6 +1394,117 @@ def _householder_product_cpu_gpu_lowering(ctx, a, taus, *,
...
+  else:
+    target_name = f"{target_name_prefix}solver_ormqr_ffi"
+  rule = _linalg_ffi_lowering(target_name, operand_output_aliases={2: 0})
+  return rule(ctx, a, taus, c, left=left, transpose=transpose)
+
+
+ormqr_p = standard_linalg_primitive(
+    (_float | _complex, _float | _complex, _float | _complex), (2, 1, 2),
+    _ormqr_shape_rule, "ormqr")
+mlir.register_lowering(ormqr_p, mlir.lower_fun(
```
Comment (hawkinsp) [inline jax/_src/lax/linalg.py:1524]: We need to make sure the Python code works with the current jaxlib release, because if nothing else our tests test that combination.

I think the easiest way to do that is to bump `_version` in `xla_client.py` and then skip adding the cpu/gpu rules if `jaxlib_extension_version` isn't at least your new version number. Then we'll fall back to your generic lowering for older jaxlibs.

File: jaxlib/gpu/solver_interface.cc [hunk 2]
```diff
@@ -131,6 +131,36 @@ JAX_GPU_DEFINE_ORGQR(gpuComplex, gpusolverDnCungqr);
...
 
+// Householder multiply: ormqr/unmqr
+
+#define JAX_GPU_DEFINE_ORMQR(Type, Name)                                       \
+  template <>                                                                  \
+  absl::StatusOr<int> OrmqrBufferSize<Type>(                                   \
+      gpusolverDnHandle_t handle, gpublasSideMode_t side,                      \
+      gpublasOperation_t trans, int m, int n, int k) {                         \
+    int lwork;                                                                 \
+    JAX_RETURN_IF_ERROR(JAX_AS_STATUS(Name##_bufferSize(                       \
```
Comment (hawkinsp) [inline jaxlib/gpu/solver_interface.cc:142]: This doesn't look right to me: https://docs.nvidia.com/cuda/cusolver/index.html#cusolverdn-t-ormqr shows this function as taking 12 parameters, you've only got 10 here. Missing C and ldc?

https://github.com/jax-ml/jax/actions/runs/22343064868/job/66848855327?pr=35104

File: jaxlib/gpu/solver_interface.cc [hunk 3]
```diff
@@ -131,6 +131,36 @@ JAX_GPU_DEFINE_ORGQR(gpuComplex, gpusolverDnCungqr);
...
+        handle, side, trans, m, n, k, /*A=*/nullptr, /*lda=*/m,                \
```
Comment (hawkinsp) [inline jaxlib/gpu/solver_interface.cc]: I'm not sure about this, but should `lda` be `n` for side == right?

File: jax/_src/lax/linalg.py [hunk 4]
```diff
@@ -1394,6 +1394,117 @@ def _householder_product_cpu_gpu_lowering(ctx, a, taus, *,
...
+  Returns:
+    The result of multiplying ``c`` by Q (or ``Q^T``/``Q^H``), with the
+    same shape as ``c``.
+  """
+  a, taus, c = core.standard_insert_pvary(a, taus, c)
+  return ormqr_p.bind(a, taus, c, left=left, transpose=transpose)
+
+
+def _ormqr_shape_rule(a_shape, taus_shape, c_shape, *, left, transpose):
+  m = a_shape[0]
```
Comment (hawkinsp) [inline jax/_src/lax/linalg.py:1450]: Do the batch dimensions have to match?

File: jax/_src/scipy/linalg.py [hunk 5]
```diff
@@ -997,6 +997,140 @@ def qr(a: ArrayLike, overwrite_a: bool = False, lwork: Any = None, mode: str = "
...
+    c = jnp.concatenate([c, zeros], axis=-2)
+
+  if conjugate:
+    # Avoid explicit jnp.conj call by instead transposing (for free under jit)
+    # and telling LAPACK to compute Hermitian
+    c = c.swapaxes(-1, -2)
+
+  # conjugate swaps left/right because c and Q change sides when transposing.
+  left = (mode == 'left') != conjugate
+  cQ = lax_linalg.ormqr(r, taus, c, left=left, transpose=conjugate)
```
Comment (hawkinsp) [inline jax/_src/scipy/linalg.py:1121]: Do you need to broadcast the batch dimension of both arrays here, if they don't match? Or do they need to match? What does scipy do here?

**Review by hawkinsp — CHANGES_REQUESTED** [review]
Sorry for the slow review, I had missed this was marked ready for review.

Please squash your commits and rebase on head.

File: jaxlib/gpu/solver_interface.cc [hunk 6]
```diff
@@ -131,6 +131,36 @@ JAX_GPU_DEFINE_ORGQR(gpuComplex, gpusolverDnCungqr);
...
+        /*tau=*/nullptr, &lwork)));                                            \
```
Comment (jpbrodrick89) [inline jaxlib/gpu/solver_interface.cc]:
```suggestion
        handle, side, trans, m, n, k, /*A=*/nullptr, /*lda=*/m,                \
        /*tau=*/nullptr, /*C=*/nullptr, /*ldc=*/m, &lwork)));                  \
```

File: jaxlib/gpu/solver_interface.cc (hunk 2 above)
Comment (jpbrodrick89) [reply jaxlib/gpu/solver_interface.cc:142]: thanks, I hadn't tested on GPU due to jaxlib rebuild time when I submitted the PR, but I created a standalone repo to test in isolation with much quicker rebuild times and spotted this bug then, merged the fix in now.

File: jaxlib/gpu/solver_interface.cc (hunk 3 above)
Comment (jpbrodrick89) [reply jaxlib/gpu/solver_interface.cc]: Yes you're completely right, good catch, this didn't actually matter in practice as CuSolver seems to just ignore this parameter but I will fix for future proofing.


```suggestion
        handle, side, trans, m, n, k, /*A=*/nullptr, /*lda=*/m,                \
        handle, side, trans, m, n, k, /*A=*/nullptr,                           \
        /*lda=*/(side == GPUBLAS_SIDE_LEFT ? m : n),                           \
```

File: jax/_src/lax/linalg.py (hunk 4 above)
Comment (jpbrodrick89) [reply jax/_src/lax/linalg.py:1450]: afaiu _ormqr_shape_rule is called by linalg_shape_rule which strips the batch dims first. Any batch dim checking should arguably be the responsibility of linalg_shape_rule. Any reason you think we should differ from the standard pattern?

Comment (hawkinsp) [reply jax/_src/lax/linalg.py:1450]: Ah yeah, you're right!

**Review by hawkinsp — COMMENTED** [review]
CUDA build is still failing.

File: jax/_src/scipy/linalg.py [hunk 7]
```diff
@@ -997,6 +997,140 @@ def qr(a: ArrayLike, overwrite_a: bool = False, lwork: Any = None, mode: str = "
...
+  if mode == 'right':
+    cQ = cQ[..., :k]
+
+  if onedim:
+    cQ = cQ.ravel()
+
+  r = jnp.triu(r[..., :k, :])
+
+  if pivoting:
+    return cQ, r, p
```
Comment (hawkinsp) [inline jax/_src/scipy/linalg.py:1136]: You'll need to do something to silence the linter about this pattern, see the `lint_and_typecheck` job. Probably just set `p=None` in the non-pivoting case above.

File: jax/_src/scipy/linalg.py (hunk 5 above)
Comment (jpbrodrick89) [reply jax/_src/scipy/linalg.py:1121]: Good catch, I've made the behaviour consistent with scipy by broadcasting dimensions now. This took a while to wrap my head around as Jax's cho_solve is not consistent with scipy's cho_solve (scipy will support a 3D batched b and unbatched c but jax.scipy.cho_solve won't), do you think this is intentional or should I file a bug report?

**Comment by jpbrodrick89** [comment]
Thanks so much for taking a look at this @hawkinsp! No worries at all about the delay, completely understandable given the volume you have to deal with. I think I've addressed everything except the squash (I have merged main), I will squash once you're happy with it, this way its easier to track any amendments I've made.

Note after looking at j-towns.github.io/papers/qr-derivative.pdf, I don't think a jvp rule for qr_multiply would be too bad, happy to add that here, but my gut feel its best to leave for a follow-up to ease cognitive load?

**Comment by jpbrodrick89** [comment]
One thing I forgot to mention was that on an A100 ormqr can be slower than dense QR for columns <500 and rows<2000 due to how well optimised gemv and how little a bottleneck memory is. For the case of columns=512 the speedup is very modest at about 10%. on CPU speedup is more universal and can reach 2.5x for 5k rows.

Should I document this at all? I feel like users would know CPU is probably better on smaller arrays anyway.

I haven't done batched benchmarks.

File: jax/_src/scipy/linalg.py (hunk 5 above)
Comment (hawkinsp) [reply jax/_src/scipy/linalg.py:1121]: I think any behavior difference between scipy and jax.scipy where it isn't simply an obvious extension of the scipy behavior (e.g., more batch dims) is a bug and please file it.

**Review by hawkinsp — CHANGES_REQUESTED** [review]
This is looking good.

You'll need to sync up again (sorry this requires updating the version number you grabbed)

and please also squash your commits: there should be only one commit showing in the "commits" tab.

**Comment by jpbrodrick89** [comment]
Fingers crossed that did it!

**Comment by jpbrodrick89** [comment]
Sorry about that missed a kwarg in the doc example, looks like everything else is passing so hopefully we're nearly there 🤞

**Review by hawkinsp — APPROVED** [review]

**Comment by hawkinsp** [comment]
Merged. Thanks for the PR and thanks for being patient with the review!
